src/OpenTelemetry.Instrumentation.ConfluentKafka/InstrumentedConsumer.cs

```diff
@@ -0,0 +1,330 @@
+    }
+
+    public List<TopicPartitionOffset> Committed(IEnumerable<TopicPartition> partitions, TimeSpan timeout)
+    {
+        return consumer.Committed(partitions, timeout);
+    }
+
+    public Offset Position(TopicPartition partition)
+    {
+        return consumer.Position(partition);
+    }
+
+    public List<TopicPartitionOffset> OffsetsForTimes(IEnumerable<TopicPartitionTimestamp> timestampsToSearch, TimeSpan timeout)
+    {
+        return consumer.OffsetsForTimes(timestampsToSearch, timeout);
+    }
+
+    public WatermarkOffsets GetWatermarkOffsets(TopicPartition topicPartition)
+    {
+        return consumer.GetWatermarkOffsets(topicPartition);
+    }
+
+    public WatermarkOffsets QueryWatermarkOffsets(TopicPartition topicPartition, TimeSpan timeout)
+    {
+        return consumer.QueryWatermarkOffsets(topicPartition, timeout);
+    }
+
+    public void Close()
+    {
+        consumer.Close();
+    }
+
+    private static PropagationContext ExtractPropagationContext(Headers headers)
+        => Propagators.DefaultTextMapPropagator.Extract(default, headers, ExtractTraceContext);
+
+    private static IEnumerable<string> ExtractTraceContext(Headers headers, string value)
+    {
+        if (headers?.TryGetLastBytes(value, out var bytes) == true)
+        {
+            yield return Encoding.UTF8.GetString(bytes);
+        }
+    }
+
+    private void InstrumentConsumption(DateTimeOffset startTime, ConsumeResult<TKey, TValue> result)
+    {
+        ConsumeResult consumeResult = new(result.TopicPartitionOffset, result.Message.Headers, result.Message.Key);
+        PropagationContext propagationContext = consumeResult.Headers != null
+            ? ExtractPropagationContext(consumeResult.Headers)
+            : default;
+        var spanName = string.IsNullOrEmpty(consumeResult.TopicPartitionOffset?.Topic)
+            ? OperationName
+            : string.Concat(consumeResult.TopicPartitionOffset!.Topic, " ", OperationName);
+
+        using var activity = this.StartReceiveActivity(spanName, propagationContext, startTime, consumeResult.TopicPartitionOffset, consumeResult.Key);
+        activity?.Stop();
+    }
+
+    private void InstrumentConsumption(DateTimeOffset startTime, ConsumeException exception)
+    {
+        ConsumeResult consumeResult = new(exception.ConsumerRecord.TopicPartitionOffset, exception.ConsumerRecord.Message.Headers, exception.ConsumerRecord.Message.Key);
+        PropagationContext propagationContext = consumeResult.Headers != null
+            ? ExtractPropagationContext(consumeResult.Headers)
+            : default;
+        var spanName = string.IsNullOrEmpty(consumeResult.TopicPartitionOffset?.Topic)
+            ? OperationName
+            : string.Concat(consumeResult.TopicPartitionOffset!.Topic, " ", OperationName);
+
+        using var activity = this.StartReceiveActivity(spanName, propagationContext, startTime, consumeResult.TopicPartitionOffset, consumeResult.Key);
+        activity?.Stop();
```

Comment:
Will this lead to useful traces?

It will show only a sliver of time for the consumption:

```
Service A (Producer) 

Incoming request _________________________
                             publish___

Service B (Consumer)

                                                  consume_
```

I feel like what users actually want is:

```
Service A (Producer) 

Incoming request _________________________
                             publish___

Service B (Consumer)

                                                  consume_________________________________
                                                             call some other service___
```

Meaning the trace started when a message is consumed should run over everything that happens processing that message, no?

Reply:
@lmolkova Any feel for how traces should be represented when it comes to message queues\pub-sub?

Reply:
We have messaging semconv with examples here https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md

The short answer - if the user calls `kafkaClient.Poll()` then we should have a receive span. If there is an API in the SDK that allows to do `kafkaClient.Process(msg => {...})`, we should have process span covering the duration of the process.
If user writes

```csharp
var msg = kafkaClient.Poll();
// process msg 
```

then there is nothing instrumentation can do to instrument things after poll call ends. Users are expected to add spans themselves.

Reply:
@g7ed6e Is there any way for the caller consuming the message to access the headers/trace information so they can continue the trace if desired?

Reply:
@CodeBlanch Yes. The caller is supposed to call `Consume` in a loop. https://docs.confluent.io/kafka-clients/dotnet/current/overview.html#the-consume-loop.
Each call to `Consume` returns a `ConsumeResult<TKey, TValue>`. Headers are here accessible through the `Message.Headers` property. https://docs.confluent.io/platform/current/clients/confluent-kafka-dotnet/_site/api/Confluent.Kafka.ConsumeResult-2.html

Reply:
@g7ed6e 

What if the instrumentation provided some extensions to help users do this correctly?

```csharp
namespace Confluent.Kafka;

public static class OpenTelemetryConsumeResultExtensions
{
    public static bool TryExtractPropagationContext<TKey, TValue>(
        this ConsumeResult<TKey, TValue> consumeResult,
        out PropagationContext propagationContext)
    {
        // todo: implement
    }

    public static void ConsumeAndProcessMessage<TKey, TValue>(
        this IConsumer<TKey, TValue> consumer,
        Action<ConsumeResult<TKey, TValue>> processMessage,
        CancellationToken cancellationToken = default)
    {
        var consumeResult = consumer.Consume(cancellationToken);

        if (TryExtractPropagationContext(consumeResult, out var propagationContext))
        {
            // todo: start child span
        }

        try
        {
            processMessage(consumeResult);
        }
        finally
        {
            // todo: stop child span
        }
    }
}
```